SConstruct

```diff
@@ -62,124 +73,149 @@ def runConstructCalib(typeString, priorStage, visitList,
-                                sourceScript='measurePhotonTransferCurve.py')
-
-env.Alias('science', science)
+ptcExposurePairs = " ".join([str(vv) for vv in exposureDict['ptcExposurePairs']])
+ptcIsrExposures = ptcExposurePairs.replace(" ", "^").replace(",", "^")
+obsLsstDir = env.ProductDir('obs_lsst')
+ptcGenIsr = env.Command(os.path.join(REPO_ROOT, 'ptcGenIsr'), flat,
+                        [getExecutableCmd('ip_isr', 'runIsr.py',
+                                          REPO_ROOT, f"--calib {CALIB_ROOT}",
+                                          "--rerun", f"{REPO_ROOT}/ptcGenIsr",
+                                          "--id detector=0",
+                                          "expId={}".format(ptcIsrExposures),
+                                          f"-C {obsLsstDir}/config/latiss/ptcIsr.py",
+                                          '-j', str(num_process))])
+env.Alias('ptcIsr', ptcGenIsr)
+
+ptcGen = env.Command(os.path.join(REPO_ROOT, 'ptcGen'), ptcGenIsr,
+                     [getExecutableCmd('cp_pipe', 'measurePhotonTransferCurve.py',
+                                       f"{REPO_ROOT}/ptcGenIsr",
+                                       "--rerun", f"{REPO_ROOT}/ptcGen",
+                                       "--id detector=0",
+                                       "-c ptcFitType=POLYNOMIAL polynomialFitDegree=3",
+                                       "--visit-pairs {}".format(ptcExposurePairs),
+                                       "-j", str(num_process))])
+env.Alias('ptcGen', ptcGen)
+
+# Brighter-Fatter Kernel.
+#    This still does ISR processing, so clip exposure list to speed processing.
+bfkExposurePairs = " ".join([str(vv) for vv in exposureDict['ptcExposurePairs'][::2]])
+bfkGen = env.Command(os.path.join(REPO_ROOT, 'bfkGen'), flat,
+                     [getExecutableCmd('cp_pipe', 'makeBrighterFatterKernel.py',
+                                       REPO_ROOT,
+                                       "--calib {CALIB_ROOT}",
+                                       "--rerun", f"{REPO_ROOT}/bfkGen",
+                                       "--id detector=0",
+                                       "--visit-pairs {}".format(bfkExposurePairs),
+                                       "-j", str(num_process))])
+env.Alias('bfkGen', bfkGen)
+
+# Tests and bookkeeping.
+everything = [butler, calibs]
+everything.extend([bias, dark, flat])
+everything.extend([bfkGen, ptcGen, defectsGen, crosstalkGen, science])
+
+tests = [env.Command(f"test_{name}", ['DATA/flatGen'],
+                     getExecutableCmd('ci_cpp_gen2', f"test_{name}.py", directory="tests"))
+         for name in ('flat', 'bias', 'dark', 'ptc', 'defects', 'crosstalk', 'brighterFatter', 'linearity')]
+env.Alias("tests", tests)
+everything.extend([tests])
+
+env.Alias("install", "SConstruct")
+
+env.Alias("all", everything)
+Default(everything)
+
+env.Clean(everything, [y for x in everything for y in x] + ['DATA'])
```

Comment:
If I understood well from the CCP standing meeting this morning, you are going to move most of the code here to call the tests to python, and then have Scons call those python scripts, correct? 
With respect to the LATISS data, I think you said you chose it because you wanted real data for the ci tests, correct? or are there any other reasons?  Or could we use some other data set that may be better behaved, such as BOT? Or even simulations?  If the tests fail, how do we know that it is not because of less-than-ideal data as opposed to mistakes in the code? 
If we wanted to change LATISS dataset in the future, would we need to change all the hard coded places in this code?

Reply:
Some of the data in the next runs that will be taken at the BOT, for example, will have datasets based on DMTN-101.

Reply:
At the time I had the idea for this, I think LATISS was the only data we had on disk (I may have been mistaken).  I chose a date that looked to have sufficient data to test most things, and set it up as the `testdata_latiss_cpp` package.  One of the suggestions that came during a sprint planning was that it would be good if the testdata is modular, although I never fully implemented it here.  At the moment, only the `TESTDATA_ROOT` value needs to be changed in the SConstruct file to point to a different product dir to change the input.  This was also the reason I moved all the visit/exposure information to the manifest.yaml file in `testdata_latiss_cpp`: so the ci and testing code is general, and all data selection choices are associated with the testdata package.

If we can get BOT data that contains everything, that would be good.  The other reason to choose LATISS is that the single chip is faster to process.  I suspect we can still do that with BOT data, by correctly restricting the dataIds.

Reply:
" At the moment, only the TESTDATA_ROOT value needs to be changed in the SConstruct file to point to a different product dir to change the input.":  I still see some specific data-id in the tests and some specific readout noise value too (I think): 2020012800014. and 0.71. Wouldn't these need to be changed too, or generalized?

Reply:
Ah, that's a good catch.  Those will need to be generalized as well.

Reply:
I don't think it's fair to say that LATISS data is not good, nor that it's any better behaved than BOT data. In many ways it's quite the opposite - at least LATISS is "nominally" finished and therefore somewhat static (though admittedly right now a little more static than anyone is happy with!).

I think the main argument for doing things with BOT data is because of it being multichip 0 some form of testing that calibration outputs don't collide, for example, or are correctly mapped to detectors when retrieved, and can be retrieved by all logically coherent combinations of detector ID, raft name/detector name etc would be good. Hopefully we can test that in a way that's robust, but only using 2 chips, otherwise it's going to be overly slow and big.